### PR TITLE
Implementing version info parse #921

### PR DESCRIPTION
**Pull request by AntonDevil**
Resource directory and version info structure parse implemented.
Parsing tested. Need to think about storing information and how
to work with UTF-16 (mb storing in sdb in UTF-8?)  Assumed that
PE file resource directory 3 levels deep (type group - entries - langs).
proposals and advices are welcome =)

### CODE REVIEW

**Comment by AntonDevil** [comment]
btw, parsing resource directory is the first step to many of PE Info #921 parsing

**Comment by XVilka** [comment]
You can store UTF-16 as a base64 in sdb.

**Comment by radare** [comment]
Yep. Use native sdb functions to do that. Anyway you can also convert utf16 to utf8. That will be more readable but you lose the original encoding, not sure if you really need that, but i prefer to work in utf8 all the time.

**Comment by Maijin** [comment]
Being able to retrieve the original content is very resources.
As you know it's quite common to find sub-dll or sub-executable as a Resources in Malware.

**Comment by radare** [comment]
The original content is inside the file. When you get a string you want it readable and accessible by the user.

**Comment by Maijin** [comment]
Yes of course if it's only string issue it's ok :+1:

File: libr/bin/format/pe/pe.c [hunk 1]
```diff
@@ -730,12 +730,648 @@ printf ("SYMBOL 0x%x = %d (%s)\n", (ut32)si->n_value, (int)si->n_strx,
...
+		bin->resource_directory = NULL;
+		return R_FALSE;
+	}
+	bin->resource_directory_offset = resource_dir_paddr;
+	return R_TRUE;
+}
+
+static void free_Var(Var *var) {
+	if (var) {
+		if (var->szKey)
```
Comment (jvoisin) [inline libr/bin/format/pe/pe.c]: No need to check if `bla` is null before calling `free (bla)` :)

File: libr/bin/format/pe/pe.c [hunk 2]
```diff
@@ -730,12 +730,648 @@ printf ("SYMBOL 0x%x = %d (%s)\n", (ut32)si->n_value, (int)si->n_strx,
...
+		free(vs_VersionInfo);
+	}
+}
+
+void PE_(free_VS_VERSIONINFO)(VS_VERSIONINFO *vs_VersionInfo) {
+	free_VS_VERSIONINFO(vs_VersionInfo);
+}
+
+static Var *Pe_r_bin_pe_parse_var(struct PE_(r_bin_pe_obj_t)* bin, PE_DWord *curAddr) {
+	Var *var = calloc (1, sizeof(*var));
```
Comment (jvoisin) [inline libr/bin/format/pe/pe.c]: What if `calloc` returns NULL?

**Comment by jvoisin** [comment]
Except the `if (bla != NULL) free (bla);` and not checking memory-allocation, the code looks pretty clean to me, that's great :)

**Comment by jvoisin** [comment]
It would be cool to have some tests to avoid regressions btw ;)

**Comment by XVilka** [comment]
This one is heavily related https://github.com/radare/radare2/issues/926

**Comment by Maijin** [comment]
@jvoisin for the moment it doesn't display anything so, it's quite hard to test. But I suppose this resource parser is launch against all PE files in format.pe tests and passes current tests.

**Comment by radare** [comment]
if you store that information in sdb, the user can access it using the `k` command using an sdb query. We can think in a generic way to interact with this information in a generic way to support other file formats later.

**Comment by AntonDevil** [comment]
Guys, you could check version info parsing. It is in bin/cur/info/vs_version_info directory in sdb. Comments, claims, complaints highly required. Thinking of adding regression test.

**Comment by AntonDevil** [comment]
Guys, you could check version info parsing. It is in bin/cur/info/vs_version_info directory in sdb. Comments, claims, complaints highly required. Thinking of adding regression test.
Hope i didn't break any tests.

**Comment by Maijin** [comment]
@AntonDevil See the green check, means that you didn't break tests ;-)

**Comment by radare** [comment]
ready to merge?

**Comment by AntonDevil** [comment]
Yes

**Comment by radare** [comment]
Those changes can't be applied to master. Can you redo the pullreq with a single-squashed commit rebased from the master branch?

thanks

**Comment by AntonDevil** [comment]
@radare done

**Comment by radare** [comment]
Thanks! merged

**Comment by radare** [comment]
Well... that broke the windows build because of dupped defines.
see http://ci.rada.re/job/radare2-w32/4108/console

**Comment by radare** [comment]
Fixed in 1aac146fdfdc9ff16812f9ba21ea9586f00ef3a6

**Pull request closed without merging**